### PR TITLE
Migrate to CLI v3

### PR DESCRIPTION
**Pull request by nebulatgs**
The repository tracking changes can be found at https://github.com/railwayapp/cliv3

**Summary of Changes**
- Migrate to using Rust with Clap for CLI args
- Use the v2 Public API
- Support for `railway service` to link to a service
- `railway domain` to generate a service domain
- Add various flags for use in CI/CD scripts

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    runs-on: ubuntu-latest
+    outputs:
+      rlwy_version: ${{ env.CLI_VERSION }}
+
+    steps:
+      - name: Get the release version from the tag
+        shell: bash
+        if: env.CLI_VERSION == ''
+        run: |
+          # Apparently, this is the right way to get a tag name. Really?
```
Comment (ndneighbor) [inline .github/workflows/release.yml:21]: I ran into the same thing a while ago. Some data model stuff for GH Actions don't make sense.

File: .github/workflows/label-check.yml [hunk 2]
```diff
@@ -11,6 +11,6 @@ jobs:
       - uses: jesusvasquez333/verify-pr-label-action@v1.4.0
         with:
           github-token: "${{ secrets.GITHUB_TOKEN }}"
-          valid-labels: "release/patch, release/minor, release/major"
-          pull-request-number: '${{ github.event.pull_request.number }}'
+          valid-labels: "release/patch, release/minor, release/major, release/skip"
```
Comment (ndneighbor) [inline .github/workflows/label-check.yml:14]: Good add! :+1:

File: .github/workflows/tag.yml [hunk 3]
```diff
@@ -1,58 +0,0 @@
-name: Tag
```
Comment (ndneighbor) [inline .github/workflows/tag.yml:1]: Slight nit: this comes from years of managing yml. I personally find .yml files unreadable whenever you have any yml file above 100 lines. There is some opportunity to maybe split the .ymls and steps, but I think this is fine for now and dependent on personal opinion.

Why do you think we should have merged the files in your opinion?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,15 +1,17 @@
...
-Build
 
-```shell
-make build
-```
+### Nix Setup
+`nix-shell` to enter the shell with all the dependencies
+
+### Running and Building
+`cargo run -- <args>` to run the binary\
```
Comment (ndneighbor) [inline CONTRIBUTING.md:16]: Trailing slash here, I assume you fat fingered that key. 😃

File: Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,74 @@
+[package]
+name = "railwayapp"
+version = "0.0.15"
```
Comment (ndneighbor) [inline Cargo.toml:3]: I assume the version would have to be set here?

File: CONTRIBUTING.md (hunk 4 above)
Comment (nebulatgs) [reply CONTRIBUTING.md:16]: The trailing slash indicates a newline in Markdown, so it's needed

File: src/commands/starship.rs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+/// Starship Metadata
+#[derive(Parser)]
+#[clap(hide = true)]
+pub struct Args {}
+
+pub async fn command(_args: Args, _json: bool) -> Result<()> {
+    let configs = Configs::new()?;
+    let linked_project = configs.get_linked_project().await?;
+    println!("{}", serde_json::to_string(&linked_project)?);
```
Comment (coffee-cup) [inline src/commands/starship.rs:11]: I don't understand how to use this command. Maybe add some descriptive docs?

image

File: uninstall.sh [hunk 7]
```diff
@@ -1,3 +0,0 @@
-rm $(which railway)
```
[the remark this answers is not in the document]
Comment (nebulatgs) [reply uninstall.sh:1]: The install script can be run with the `-r` flag to uninstall

File: src/commands/starship.rs (hunk 6 above)
Comment (nebulatgs) [reply src/commands/starship.rs:11]: This is a hidden command for `starship` (https://github.com/starship/starship)
It's used to get metadata for the prompt - it doesn't appear in help and is not meant for direct use by people.

File: Cargo.toml (hunk 5 above)
Comment (nebulatgs) [reply Cargo.toml:3]: Cargo release should set this once we release it to crates.io

File: src/commands/domain.rs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        bail!("No services found for project");
+    }
+
+    // If there is only one service, it will generate a domain for that service
+    let service = if body.project.services.edges.len() == 1 {
+        body.project.services.edges[0].node.clone().id
+    } else {
+        let Some(service) = linked_project.service.clone() else {
+        bail!("No service linked");
+      };
```
[the remark this answers is not in the document]
Comment (nebulatgs) [reply src/commands/domain.rs]: Added a check for this

File: src/commands/domain.rs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        post_graphql::<queries::ServiceDomains, _>(&client, configs.get_backboard(), vars).await?;
+
+    let body = res
+        .data
+        .context("Failed to retrieve to get domains for service.")?;
+
+    let domain = body.domains;
+    if domain.service_domains.is_empty() || domain.custom_domains.is_empty() {
+        bail!("Domain already exists on service");
+    }
```
[the remark this answers is not in the document]
Comment (nebulatgs) [reply src/commands/domain.rs]: Fixed by negating

File: src/commands/add.rs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        if filtered.is_empty() {
+            bail!("Plugins already exist");
+        }
+
+        filtered
+    } else {
+        inquire::MultiSelect::new("Select plugins to add", filtered_plugins)
+            .with_render_config(render_config)
+            .prompt()?
+    };
```
[the remark this answers is not in the document]
Comment (JakeCooper) [reply src/commands/add.rs:70]: When we can, please try to keep code linear

Logic trees should be avoided if early returns will do (which they usually will)

File: src/commands/domain.rs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    let body = res.data.context("Failed to retrieve response body")?;
+
+    if body.project.services.edges.is_empty() {
+        bail!("No services found for project");
+    }
+
+    // If there is only one service, it will generate a domain for that service
+    let service = if body.project.services.edges.len() == 1 {
+        body.project.services.edges[0].node.clone().id
+    } else {
```
Comment (JakeCooper) [inline src/commands/domain.rs:36]: Does line 35 return? If so, we should be using explicit returns

Implicit returns are fine for small functions like map/filter/return but should be avoided, for clarity, here

Also, if the else is going to bail, we should flip the branches and have an early return to remove the else case

File: src/config.rs [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+        let path = self.get_closest_linked_project_directory()?;
+        let project = self
+            .root_config
+            .projects
+            .get(&path)
+            .context("Project not found! Run `railway link` to link to a project")?;
+        Ok(project.clone())
+    }
+
+    pub fn get_linked_project_mut(&mut self) -> Result<&mut RailwayProject> {
```
[the remark this answers is not in the document]
Comment (nebulatgs) [reply src/config.rs:202]: This is for mutating the project specified in the configuration file, project tokens work without touching the config file, so we don't need to handle them here

File: src/consts.rs [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+pub const TICK_STRING: &str = "⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏ ";
+
+pub const PLUGINS: &[&str] = &["PostgreSQL", "MySQL", "Redis", "MongoDB"];
+
+pub const NO_SERVICE_LINKED: &str =
+    "No service linked and no plugins found\nRun `railway service` to link a service";
+pub const ABORTED_BY_USER: &str = "Aborted by user";
+pub const PROJECT_NOT_FOUND: &str = "Project not found!";
+pub const SERVICE_NOT_FOUND: &str = "Service not found!";
+pub const NON_INTERACTIVE_FAILURE: &str = "This command is only available in interactive mode";
```
[the remark this answers is not in the document]
Comment (nebulatgs) [reply src/consts.rs:14]: Not sure, is that more idiomatic?

File: src/commands/link.rs [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+                body.project
+                    .environments
+                    .edges
+                    .iter()
+                    .map(|env| ProjectEnvironment(&env.node))
+                    .collect(),
+            )
+            .with_render_config(configs.get_render_config())
+            .prompt()?
+        };
```
Comment (JakeCooper) [inline src/commands/link.rs:70]: Logic that probably would benefit from being pulled into a function

**Review by JakeCooper — COMMENTED** [review]
I'd say in general we need to take a pass over this to make it more clear and clean

I know you're trying to land this for today, so we can just fix the flags I've added for now, but we're gonna want to go over this again and iron it out because IMO, there are a few places for which it's a bit awkward

File: src/consts.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (coffee-cup) [reply src/consts.rs:14]: I think would be a little bit. But very low priority in the grand scheme of things

File: src/commands/add.rs [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+                        .tick_chars(TICK_STRING)
+                        .template("{spinner:.green} {msg}")?,
+                )
+                .with_message(format!("Creating {plugin}..."));
+            spinner.enable_steady_tick(Duration::from_millis(100));
+            post_graphql::<mutations::PluginCreate, _>(&client, configs.get_backboard(), vars)
+                .await?;
+            spinner.finish_with_message(format!("Created {plugin}"));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (coffee-cup) [reply src/commands/add.rs:98]: Or just abstract it ourselves. We shouldn't have to duplicate the gql query every time depending on if we are in a terminal environment or not

File: src/commands/add.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (nebulatgs) [reply src/commands/add.rs:70]: This isn't an early return, it's a conditional value (similar to a ternary expression in other languages), so the else is needed

File: src/commands/domain.rs (hunk 11 above)
Comment (nebulatgs) [reply src/commands/domain.rs:36]: This another instance of a conditional value

File: src/commands/init.rs [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+        .with_render_config(render_config)
+        .prompt()?;
+
+    let name = if name.is_empty() {
+        use names::Generator;
+        let mut generator = Generator::default();
+        generator.next().context("Failed to generate name")?
+    } else {
+        name
+    };
```
[the remark this answers is not in the document]
Comment (coffee-cup) [reply src/commands/init.rs:46]: Agreed. In general I think we should abstract out the `inquire` library so we can do things like

```rs
let name = prompt("Project Name", "{placeholder}", "{help message}").or_else(|| random_name())
```

or some variation of that. Basically condensing these lines

[the remark this answers is not in the document]
Comment (nebulatgs) [reply src/commands/init.rs:46]: I think this is a great idea, will do!

**Review by ndneighbor — APPROVED** [review]
Green, expected fast follow on for code quality.

**Pull request merged**